File: backend/pdf_processor.py
import pdfplumber
import pandas as pd
import re
from typing import List, Dict, Any, Optional, Tuple
from datetime import datetime
import logging
from io import StringIO
import numpy as np

logger = logging.getLogger(__name__)

class PDFBankStatementProcessor:
    def __init__(self):
        self.bank_patterns = {
            'sbi': ['state bank of india', 'sbi', 'state bank'],
            'hdfc': ['hdfc bank', 'hdfc', 'housing development finance'],
            'icici': ['icici bank', 'icici', 'industrial credit'],
            'axis': ['axis bank', 'axis'],
            'pnb': ['punjab national bank', 'pnb'],
            'kotak': ['kotak mahindra bank', 'kotak'],
            'indusind': ['indusind bank', 'indusind'],
            'yes': ['yes bank', 'yes'],
            'bob': ['bank of baroda', 'baroda'],
            'canara': ['canara bank', 'canara'],
            'union': ['union bank', 'union'],
            'indian': ['indian bank', 'indian'],
            'central': ['central bank', 'central'],
            'idbi': ['idbi bank', 'idbi'],
            'idfc': ['idfc first bank', 'idfc']
        }
        
        self.date_patterns = [
            r'\d{1,2}[-/]\d{1,2}[-/]\d{2,4}',
            r'\d{1,2}[-/][A-Za-z]{3}[-/]\d{2,4}',
            r'\d{2,4}[-/]\d{1,2}[-/]\d{1,2}',
            r'\d{1,2}\s+[A-Za-z]{3}\s+\d{2,4}',
            r'\d{1,2}-[A-Za-z]{3}-\d{2,4}'
        ]
        
        self.amount_patterns = [
            r'[\d,]+\.?\d*',
            r'\d+\.?\d*',
            r'[\d,]+\.\d{2}'
        ]

    def detect_bank(self, text: str) -> Optional[str]:
        """Detect bank from PDF text content"""
        text_lower = text.lower()
        
        for bank_code, patterns in self.bank_patterns.items():
            for pattern in patterns:
                if pattern in text_lower:
                    return bank_code
        return None

    def extract_tables_from_pdf(self, pdf_path: str) -> List[pd.DataFrame]:
        """Extract all potential transaction tables from PDF"""
        tables = []
        
        try:
            with pdfplumber.open(pdf_path) as pdf:
                for page_num, page in enumerate(pdf.pages):
                    logger.info(f"Processing page {page_num + 1}")
                    
                    # Try to extract tables using pdfplumber
                    page_tables = page.extract_tables()
                    
                    for table in page_tables:
                        if table and len(table) > 1:  # Must have header + at least one row
                            df = pd.DataFrame(table[1:], columns=table[0])
                            # Clean empty columns and rows
                            df = df.dropna(how='all', axis=0).dropna(how='all', axis=1)
                            if not df.empty and len(df) > 0:
                                tables.append(df)
                    
                    # If no tables found, try text extraction and parsing
                    if not page_tables:
                        text = page.extract_text()
                        if text:
                            parsed_table = self.parse_text_to_table(text)
                            if parsed_table is not None:
                                tables.append(parsed_table)
                                
        except Exception as e:
            logger.error(f"Error extracting tables: {str(e)}")
            
        return tables

    def parse_text_to_table(self, text: str) -> Optional[pd.DataFrame]:
        """Parse structured text into table format"""
        lines = text.split('\n')
        potential_rows = []
        
        for line in lines:
            line = line.strip()
            if not line:
                continue
                
            # Look for lines that might contain transaction data
            if self.is_potential_transaction_line(line):
                parsed_row = self.parse_transaction_line(line)
                if parsed_row:
                    potential_rows.append(parsed_row)
        
        if potential_rows and len(potential_rows) > 0:
            df = pd.DataFrame(potential_rows)
            return df
        
        return None

    def is_potential_transaction_line(self, line: str) -> bool:
        """Check if line contains potential transaction data"""
        # Look for date patterns
        date_found = any(re.search(pattern, line) for pattern in self.date_patterns)
        
        # Look for amount patterns
        amount_found = any(re.search(pattern, line) for pattern in self.amount_patterns)
        
        # Check for minimum length and contains both date and amount
        return len(line) > 20 and date_found and amount_found

    def parse_transaction_line(self, line: str) -> Optional[Dict[str, str]]:
        """Parse a single transaction line into structured data"""
        try:
            parts = re.split(r'\s{2,}', line.strip())  # Split on multiple spaces
            
            if len(parts) < 3:
                return None
            
            # Try to identify date, description, and amounts
            date_part = None
            amounts = []
            description_parts = []
            
            for part in parts:
                part = part.strip()
                if not part:
                    continue
                    
                # Check if it's a date
                if any(re.match(pattern, part) for pattern in self.date_patterns):
                    date_part = part
                # Check if it's an amount
                elif re.match(r'[\d,]+\.?\d*$', part.replace(',', '')):
                    amounts.append(part)
                else:
                    description_parts.append(part)
            
            if date_part and amounts:
                return {
                    'date': date_part,
                    'description': ' '.join(description_parts),
                    'amount_1': amounts[0] if len(amounts) > 0 else '',
                    'amount_2': amounts[1] if len(amounts) > 1 else '',
                    'amount_3': amounts[2] if len(amounts) > 2 else ''
                }
                
        except Exception as e:
            logger.error(f"Error parsing transaction line: {str(e)}")
        
        return None

    def identify_transaction_columns(self, df: pd.DataFrame) -> Dict[str, str]:
        """Identify which columns contain transaction data"""
        column_mapping = {}
        
        # Common column name patterns
        date_patterns = ['date', 'txn date', 'transaction date', 'value date', 'posting date']
        desc_patterns = ['description', 'particulars', 'details', 'narration', 'transaction details']
        debit_patterns = ['debit', 'withdrawal', 'dr', 'paid', 'debits']
        credit_patterns = ['credit', 'deposit', 'cr', 'received', 'credits']
        balance_patterns = ['balance', 'closing balance', 'running balance', 'available balance']
        
        for col in df.columns:
            if col is None:
                continue
                
            col_lower = str(col).lower().strip()
            
            # Match date column
            if any(pattern in col_lower for pattern in date_patterns) and 'date' not in column_mapping:
                column_mapping['date'] = col
            
            # Match description column
            elif any(pattern in col_lower for pattern in desc_patterns) and 'description' not in column_mapping:
                column_mapping['description'] = col
            
            # Match debit column
            elif any(pattern in col_lower for pattern in debit_patterns) and 'debit' not in column_mapping:
                column_mapping['debit'] = col
            
            # Match credit column
            elif any(pattern in col_lower for pattern in credit_patterns) and 'credit' not in column_mapping:
                column_mapping['credit'] = col
            
            # Match balance column
            elif any(pattern in col_lower for pattern in balance_patterns) and 'balance' not in column_mapping:
                column_mapping['balance'] = col
        
        return column_mapping

    def clean_amount(self, amount_str: str) -> float:
        """Clean and convert amount string to float"""
        if pd.isna(amount_str) or amount_str == '' or amount_str is None:
            return 0.0
        
        # Remove currency symbols and extra spaces
        amount_str = str(amount_str).strip()
        amount_str = re.sub(r'[₹$,\s]', '', amount_str)
        
        # Handle parentheses for negative amounts
        if amount_str.startswith('(') and amount_str.endswith(')'):
            amount_str = amount_str[1:-1]
            return -float(amount_str) if amount_str else 0.0
        
        try:
            return float(amount_str) if amount_str else 0.0
        except ValueError:
            return 0.0

    def normalize_date(self, date_str: str) -> str:
        """Normalize date to DD-MM-YYYY format"""
        if pd.isna(date_str) or not date_str:
            return ''
        
        date_str = str(date_str).strip()
        
        # Common date formats to try
        formats = [
            '%d-%m-%Y', '%d/%m/%Y', '%d.%m.%Y',
            '%d-%m-%y', '%d/%m/%y', '%d.%m.%y',
            '%d-%b-%Y', '%d-%b-%y',
            '%d %b %Y', '%d %b %y',
            '%Y-%m-%d', '%Y/%m/%d'
        ]
        
        for fmt in formats:
            try:
                parsed_date = datetime.strptime(date_str, fmt)
                return parsed_date.strftime('%d-%m-%Y')
            except ValueError:
                continue
        
        return date_str  # Return original if can't parse

    def process_transactions(self, tables: List[pd.DataFrame]) -> List[Dict[str, Any]]:
        """Process extracted tables into standardized transaction format"""
        all_transactions = []
        
        for table_idx, df in enumerate(tables):
            logger.info(f"Processing table {table_idx + 1} with {len(df)} rows")
            
            if df.empty:
                continue
            
            # Identify column mappings
            column_mapping = self.identify_transaction_columns(df)
            logger.info(f"Column mapping: {column_mapping}")
            
            # If we couldn't map essential columns, try alternate approach
            if not column_mapping.get('date') or not column_mapping.get('description'):
                # Try to process using positional logic
                transactions = self.process_table_positional(df)
                all_transactions.extend(transactions)
                continue
            
            # Process mapped columns
            for idx, row in df.iterrows():
                try:
                    transaction = self.extract_transaction_from_row(row, column_mapping)
                    if transaction:
                        all_transactions.append(transaction)
                except Exception as e:
                    logger.error(f"Error processing row {idx}: {str(e)}")
                    continue
        
        return all_transactions

    def process_table_positional(self, df: pd.DataFrame) -> List[Dict[str, Any]]:
        """Process table using positional logic when column mapping fails"""
        transactions = []
        
        for idx, row in df.iterrows():
            try:
                row_values = [str(val).strip() if val is not None else '' for val in row.values]
                
                # Skip empty rows
                if all(val == '' for val in row_values):
                    continue
                
                # Try to find date in first few columns
                date_col = None
                for i, val in enumerate(row_values[:3]):
                    if any(re.search(pattern, val) for pattern in self.date_patterns):
                        date_col = i
                        break
                
                if date_col is None:
                    continue
                
                # Extract transaction data
                date_val = self.normalize_date(row_values[date_col])
                
                # Find description (usually next non-amount column after date)
                description = ''
                amounts = []
                
                for i, val in enumerate(row_values):
                    if i == date_col:
                        continue
                    
                    # Check if it looks like an amount
                    if re.match(r'[\d,]+\.?\d*$', val.replace(',', '')) and val != '':
                        amounts.append(self.clean_amount(val))
                    elif val != '' and description == '':
                        description = val
                
                if date_val and description and amounts:
                    # Determine transaction type and amount
                    if len(amounts) == 1:
                        transaction_type = 'Debit' if amounts[0] > 0 else 'Credit'
                        amount = abs(amounts[0])
                    elif len(amounts) >= 2:
                        # Usually debit, credit, balance format
                        debit_amt = amounts[0] if amounts[0] > 0 else 0
                        credit_amt = amounts[1] if len(amounts) > 1 and amounts[1] > 0 else 0
                        
                        if debit_amt > 0:
                            transaction_type = 'Debit'
                            amount = debit_amt
                        elif credit_amt > 0:
                            transaction_type = 'Credit'
                            amount = credit_amt
                        else:
                            continue
                    else:
                        continue
                    
                    transaction = {
                        'date': date_val,
                        'description': description,
                        'type': transaction_type,
                        'amount': amount,
                        'balance': amounts[-1] if len(amounts) > 2 else None,
                        'reference': ''
                    }
                    
                    transactions.append(transaction)
                    
            except Exception as e:
                logger.error(f"Error in positional processing row {idx}: {str(e)}")
                continue
        
        return transactions

    def extract_transaction_from_row(self, row: pd.Series, column_mapping: Dict[str, str]) -> Optional[Dict[str, Any]]:
        """Extract transaction data from a table row"""
        try:
            date_val = self.normalize_date(row.get(column_mapping.get('date', ''), ''))
            description = str(row.get(column_mapping.get('description', ''), '')).strip()
            
            if not date_val or not description:
                return None
            
            # Handle debit/credit columns
            debit_amount = self.clean_amount(row.get(column_mapping.get('debit', ''), 0))
            credit_amount = self.clean_amount(row.get(column_mapping.get('credit', ''), 0))
            balance = self.clean_amount(row.get(column_mapping.get('balance', ''), 0))
            
            # Determine transaction type and amount
            if debit_amount > 0:
                transaction_type = 'Debit'
                amount = debit_amount
            elif credit_amount > 0:
                transaction_type = 'Credit'
                amount = credit_amount
            else:
                # Try to find amount in other columns
                for col_name, value in row.items():
                    if col_name not in column_mapping.values():
                        cleaned_amount = self.clean_amount(value)
                        if cleaned_amount > 0:
                            transaction_type = 'Debit'  # Default assumption
                            amount = cleaned_amount
                            break
                else:
                    return None
            
            return {
                'date': date_val,
                'description': description,
                'type': transaction_type,
                'amount': amount,
                'balance': balance if balance > 0 else None,
                'reference': ''
            }
            
        except Exception as e:
            logger.error(f"Error extracting transaction: {str(e)}")
            return None

    def process_pdf(self, pdf_path: str) -> Dict[str, Any]:
        """Main method to process PDF and extract transactions"""
        try:
            # Extract text for bank detection
            with pdfplumber.open(pdf_path) as pdf:
                full_text = ''
                for page in pdf.pages:
                    page_text = page.extract_text()
                    if page_text:
                        full_text += page_text + '\n'
            
            # Detect bank
            detected_bank = self.detect_bank(full_text)
            
            # Extract tables
            tables = self.extract_tables_from_pdf(pdf_path)
            
            if not tables:
                return {
                    'success': False,
                    'error': 'No transaction tables found in PDF',
                    'transactions': [],
                    'bank_name': detected_bank
                }
            
            # Process transactions
            transactions = self.process_transactions(tables)
            
            if not transactions:
                return {
                    'success': False,
                    'error': 'No valid transactions found',
                    'transactions': [],
                    'bank_name': detected_bank
                }
            
            # Sort transactions by date
            transactions.sort(key=lambda x: datetime.strptime(x['date'], '%d-%m-%Y') if x['date'] else datetime.min)
            
            # Calculate date range
            dates = [t['date'] for t in transactions if t['date']]
            date_range = None
            if dates:
                date_range = {
                    'start': dates[0],
                    'end': dates[-1]
                }
            
            return {
                'success': True,
                'transactions': transactions,
                'bank_name': self.get_bank_display_name(detected_bank),
                'transaction_count': len(transactions),
                'date_range': date_range
            }
            
        except Exception as e:
            logger.error(f"Error processing PDF: {str(e)}")
            return {
                'success': False,
                'error': f'Error processing PDF: {str(e)}',
                'transactions': [],
                'bank_name': None
            }

    def get_bank_display_name(self, bank_code: Optional[str]) -> str:
        """Get display name for bank code"""
        bank_names = {
            'sbi': 'State Bank of India',
            'hdfc': 'HDFC Bank',
            'icici': 'ICICI Bank',
            'axis': 'Axis Bank',
            'pnb': 'Punjab National Bank',
            'kotak': 'Kotak Mahindra Bank',
            'indusind': 'IndusInd Bank',
            'yes': 'Yes Bank',
            'bob': 'Bank of Baroda',
            'canara': 'Canara Bank',
            'union': 'Union Bank of India',
            'indian': 'Indian Bank',
            'central': 'Central Bank of India',
            'idbi': 'IDBI Bank',
            'idfc': 'IDFC First Bank'
        }
        
        return bank_names.get(bank_code, 'Unknown Bank') if bank_code else 'Unknown Bank'

File: src/utils/pdfProcessor.ts
import * as fs from 'fs';
import PDFParser from 'pdf2json';
import { Transaction } from '@/types';

interface BankPatterns {
  [key: string]: string[];
}

interface ProcessedResult {
  success: boolean;
  transactions: Transaction[];
  bank_name: string;
  transaction_count: number;
  date_range?: {
    start: string;
    end: string;
  };
  error?: string;
}

export class PDFBankStatementProcessor {
  private bankPatterns: BankPatterns = {
    sbi: ['state bank of india', 'sbi', 'state bank'],
    hdfc: ['hdfc bank', 'hdfc', 'housing development finance'],
    icici: ['icici bank', 'icici', 'industrial credit'],
    axis: ['axis bank', 'axis'],
    pnb: ['punjab national bank', 'pnb'],
    kotak: ['kotak mahindra bank', 'kotak'],
    indusind: ['indusind bank', 'indusind'],
    yes: ['yes bank', 'yes'],
    bob: ['bank of baroda', 'baroda'],
    canara: ['canara bank', 'canara'],
    union: ['union bank', 'union'],
    indian: ['indian bank', 'indian'],
    central: ['central bank', 'central'],
    idbi: ['idbi bank', 'idbi'],
    idfc: ['idfc first bank', 'idfc']
  };

  private datePatterns = [
    /\d{1,2}[-/]\d{1,2}[-/]\d{2,4}/g,
    /\d{1,2}[-/][A-Za-z]{3}[-/]\d{2,4}/g,
    /\d{2,4}[-/]\d{1,2}[-/]\d{1,2}/g,
    /\d{1,2}\s+[A-Za-z]{3}\s+\d{2,4}/g,
    /\d{1,2}-[A-Za-z]{3}-\d{2,4}/g
  ];

  private amountPatterns = [
    /[\d,]+\.?\d*/g,
    /\d+\.?\d*/g,
    /[\d,]+\.\d{2}/g
  ];

  private detectBank(text: string): string | null {
    const textLower = text.toLowerCase();
    
    for (const [bankCode, patterns] of Object.entries(this.bankPatterns)) {
      for (const pattern of patterns) {
        if (textLower.includes(pattern)) {
          return bankCode;
        }
      }
    }
    return null;
  }

  private getBankDisplayName(bankCode: string | null): string {
    const bankNames: { [key: string]: string } = {
      sbi: 'State Bank of India',
      hdfc: 'HDFC Bank',
      icici: 'ICICI Bank',
      axis: 'Axis Bank',
      pnb: 'Punjab National Bank',
      kotak: 'Kotak Mahindra Bank',
      indusind: 'IndusInd Bank',
      yes: 'Yes Bank',
      bob: 'Bank of Baroda',
      canara: 'Canara Bank',
      union: 'Union Bank of India',
      indian: 'Indian Bank',
      central: 'Central Bank of India',
      idbi: 'IDBI Bank',
      idfc: 'IDFC First Bank'
    };
    
    return bankNames[bankCode || ''] || 'Unknown Bank';
  }

  private isPotentialTransactionLine(line: string): boolean {
    // Look for date patterns
    const dateFound = this.datePatterns.some(pattern => pattern.test(line));
    
    // Look for amount patterns (more flexible)
    const amountFound = /\d+[.,]\d{2}|\d{4,}/.test(line); // Find amounts with decimals or large numbers
    
    // More flexible criteria - just need date OR amount with reasonable length
    const hasDate = dateFound;
    const hasAmount = amountFound;
    const hasLength = line.length > 10; // Reduced minimum length
    
    // Transaction keywords that might indicate a transaction line
    const transactionKeywords = /debit|credit|withdrawal|deposit|transfer|payment|upi|neft|rtgs|imps|cheque|cash|atm/i;
    const hasKeyword = transactionKeywords.test(line);
    
    // Accept line if it has date, OR amount with keywords, OR both date and amount
    return hasLength && (
      (hasDate && hasAmount) ||  // Ideal case
      (hasDate && hasKeyword) || // Date with transaction keyword
      (hasAmount && hasKeyword) || // Amount with transaction keyword
      (hasDate && line.split(/\s+/).length >= 4) // Date with multiple parts
    );
  }

  private parseTransactionLine(line: string): Partial<Transaction> | null {
    try {
      // Try multiple splitting strategies
      let parts = line.split(/\s{2,}/); // Split on multiple spaces
      
      // If that doesn't work well, try single space split for dense layouts
      if (parts.length < 3) {
        parts = line.split(/\s+/).filter(part => part.trim().length > 0);
      }
      
      if (parts.length < 2) {
        return null;
      }
      
      // Try to identify date, description, and amounts
      let datePart: string | null = null;
      const amounts: string[] = [];
      const descriptionParts: string[] = [];
      
      for (const part of parts) {
        const trimmedPart = part.trim();
        if (!trimmedPart) continue;
        
        // Check if it's a date
        if (this.datePatterns.some(pattern => pattern.test(trimmedPart))) {
          datePart = trimmedPart;
        }
        // Check if it's an amount (more flexible patterns)
        else if (/^[\d,]+\.?\d*$/.test(trimmedPart.replace(',', '')) && 
                 parseFloat(trimmedPart.replace(',', '')) > 0) {
          amounts.push(trimmedPart);
        }
        // Check for amounts with currency symbols
        else if (/^[₹$]?[\d,]+\.?\d*$/.test(trimmedPart.replace(',', ''))) {
          amounts.push(trimmedPart.replace(/[₹$]/, ''));
        }
        // Check for negative amounts in parentheses  
        else if (/^\([\d,]+\.?\d*\)$/.test(trimmedPart)) {
          amounts.push(trimmedPart);
        } else {
          descriptionParts.push(trimmedPart);
        }
      }
      
      // If we found a date, try to extract transaction info
      if (datePart || amounts.length > 0) {
        const description = descriptionParts.join(' ').trim();
        
        // If no explicit date found, try to find it in the description
        if (!datePart && description) {
          const dateMatch = description.match(/\b\d{1,2}[-/.]\d{1,2}[-/.]\d{2,4}\b/);
          if (dateMatch) {
            datePart = dateMatch[0];
            // Remove date from description
            descriptionParts.forEach((part, index) => {
              if (part.includes(dateMatch[0])) {
                descriptionParts[index] = part.replace(dateMatch[0], '').trim();
              }
            });
          }
        }
        
        if (datePart || (amounts.length > 0 && description)) {
          return {
            date: datePart ? this.normalizeDate(datePart) : '',
            description: descriptionParts.join(' ').trim() || 'Transaction',
            amount: amounts.length > 0 ? this.cleanAmount(amounts[0]) : 0,
            balance: amounts.length > 1 ? this.cleanAmount(amounts[amounts.length - 1]) : undefined,
            reference: ''
          };
        }
      }
    } catch (error) {
      console.error('Error parsing transaction line:', error);
    }
    
    return null;
  }

  private cleanAmount(amountStr: string): number {
    if (!amountStr || amountStr === '') {
      return 0.0;
    }
    
    // Remove currency symbols and extra spaces
    let cleanStr = amountStr.replace(/[₹$,\s]/g, '');
    
    // Handle parentheses for negative amounts
    if (cleanStr.startsWith('(') && cleanStr.endsWith(')')) {
      cleanStr = cleanStr.slice(1, -1);
      return -parseFloat(cleanStr) || 0.0;
    }
    
    return parseFloat(cleanStr) || 0.0;
  }

  private normalizeDate(dateStr: string): string {
    if (!dateStr) {
      return '';
    }
    
    dateStr = dateStr.trim();
    
    // Common date formats to try
    const formats = [
      { pattern: /(\d{1,2})[-/](\d{1,2})[-/](\d{4})/, format: 'DD-MM-YYYY' },
      { pattern: /(\d{1,2})[-/](\d{1,2})[-/](\d{2})/, format: 'DD-MM-YY' },
      { pattern: /(\d{4})[-/](\d{1,2})[-/](\d{1,2})/, format: 'YYYY-MM-DD' },
    ];
    
    for (const { pattern, format } of formats) {
      const match = dateStr.match(pattern);
      if (match) {
        if (format === 'DD-MM-YYYY') {
          return `${match[1].padStart(2, '0')}-${match[2].padStart(2, '0')}-${match[3]}`;
        } else if (format === 'DD-MM-YY') {
          const year = parseInt(match[3]) < 50 ? `20${match[3]}` : `19${match[3]}`;
          return `${match[1].padStart(2, '0')}-${match[2].padStart(2, '0')}-${year}`;
        } else if (format === 'YYYY-MM-DD') {
          return `${match[3].padStart(2, '0')}-${match[2].padStart(2, '0')}-${match[1]}`;
        }
      }
    }
    
    return dateStr; // Return original if can't parse
  }

  private parseTextToTransactions(text: string): Transaction[] {
    const lines = text.split('\n');
    const transactions: Transaction[] = [];
    
    // Also try processing by combining adjacent lines (in case transactions span multiple lines)
    const combinedLines: string[] = [];
    for (let i = 0; i < lines.length; i++) {
      const currentLine = lines[i].trim();
      if (!currentLine) continue;
      
      // Try current line alone
      combinedLines.push(currentLine);
      
      // Try combining with next line if it doesn't look complete
      if (i + 1 < lines.length) {
        const nextLine = lines[i + 1].trim();
        if (nextLine && !this.datePatterns.some(pattern => pattern.test(nextLine))) {
          combinedLines.push(currentLine + ' ' + nextLine);
        }
      }
    }
    
    // Process all lines (original and combined)
    const allLinesToProcess = [...lines.map(l => l.trim()), ...combinedLines];
    
    for (const line of allLinesToProcess) {
      if (!line) continue;
      
      if (this.isPotentialTransactionLine(line)) {
        const parsedTransaction = this.parseTransactionLine(line);
        if (parsedTransaction) {
          // More flexible validation - accept if we have amount or reasonable description
          const hasValidDate = parsedTransaction.date && parsedTransaction.date.length > 0;
          const hasValidDescription = parsedTransaction.description && parsedTransaction.description.length > 2;
          const hasValidAmount = parsedTransaction.amount !== undefined && parsedTransaction.amount !== 0;
          
          if ((hasValidDate && hasValidAmount) || (hasValidDescription && hasValidAmount)) {
            // Smart transaction type detection
            let transactionType: 'Credit' | 'Debit' = 'Debit';
            
            const desc = parsedTransaction.description?.toLowerCase() || '';
            const creditKeywords = /credit|deposit|salary|interest|refund|cashback|transfer.*credit/i;
            const debitKeywords = /debit|withdrawal|payment|purchase|charge|fee|transfer.*debit|atm/i;
            
            if (creditKeywords.test(desc)) {
              transactionType = 'Credit';
            } else if (debitKeywords.test(desc)) {
              transactionType = 'Debit';
            } else {
              // Default based on amount being positive or negative
              transactionType = (parsedTransaction.amount || 0) > 0 ? 'Credit' : 'Debit';
            }
            
            const transaction: Transaction = {
              date: parsedTransaction.date || new Date().toISOString().split('T')[0],
              description: parsedTransaction.description || 'Transaction',
              type: transactionType,
              amount: Math.abs(parsedTransaction.amount || 0),
              balance: parsedTransaction.balance || null,
              reference: parsedTransaction.reference || null
            };
            
            // Avoid duplicates
            const isDuplicate = transactions.some(existing => 
              existing.date === transaction.date &&
              existing.description === transaction.description &&
              existing.amount === transaction.amount
            );
            
            if (!isDuplicate && transaction.amount > 0) {
              transactions.push(transaction);
            }
          }
        }
      }
    }
    
    return transactions;
  }

  public async processPDF(pdfPath: string): Promise<ProcessedResult> {
    try {
      // Use pdf2json to parse PDF
      const pdfParser = new PDFParser();
      
      let fullText = '';
      
      // Create promise to handle PDF parsing
      const pdfText = await new Promise<string>((resolve, reject) => {
        pdfParser.on('pdfParser_dataError', reject);
        pdfParser.on('pdfParser_dataReady', (pdfData: any) => {
          let text = '';
          
          // Extract text from all pages
          if (pdfData && pdfData.Pages) {
            for (const page of pdfData.Pages) {
              if (page.Texts) {
                for (const textBlock of page.Texts) {
                  if (textBlock.R && textBlock.R.length > 0) {
                    const textContent = textBlock.R[0].T;
                    if (textContent) {
                      // Decode URI component to handle special characters
                      text += decodeURIComponent(textContent) + ' ';
                    }
                  }
                }
              }
              text += '\n';
            }
          }
          
          resolve(text);
        });
        
        pdfParser.loadPDF(pdfPath);
      });
      
      fullText = pdfText;
      
      // Detect bank
      const detectedBank = this.detectBank(fullText);
      
      // Extract transactions from text
      const transactions = this.parseTextToTransactions(fullText);
      
      if (transactions.length === 0) {
        return {
          success: false,
          error: 'No valid transactions found in PDF',
          transactions: [],
          bank_name: this.getBankDisplayName(detectedBank),
          transaction_count: 0
        };
      }
      
      // Sort transactions by date
      transactions.sort((a, b) => {
        const dateA = new Date(a.date.split('-').reverse().join('-'));
        const dateB = new Date(b.date.split('-').reverse().join('-'));
        return dateA.getTime() - dateB.getTime();
      });
      
      // Calculate date range
      const dates = transactions.map(t => t.date).filter(Boolean);
      let dateRange;
      if (dates.length > 0) {
        dateRange = {
          start: dates[0],
          end: dates[dates.length - 1]
        };
      }
      
      return {
        success: true,
        transactions,
        bank_name: this.getBankDisplayName(detectedBank),
        transaction_count: transactions.length,
        date_range: dateRange
      };
      
    } catch (error) {
      console.error('Error processing PDF:', error);
      return {
        success: false,
        error: `Error processing PDF: ${error}`,
        transactions: [],
        bank_name: 'Unknown Bank',
        transaction_count: 0
      };
    }
  }
}